### PR TITLE
test(bigquerycontrol): add job integration tests

### PR DESCRIPTION
**Pull request by scotthart**
part of the work for #14456

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14457)

### CODE REVIEW

File: google/cloud/bigquerycontrol/CMakeLists.txt [hunk 1]
```diff
@@ -197,6 +197,10 @@ if (BUILD_TESTING AND GOOGLE_CLOUD_CPP_ENABLE_CXX_EXCEPTIONS)
                          PROPERTIES LABELS "integration-test;quickstart")
 endif ()
 
+if (BUILD_TESTING)
```
Comment (coryan) [inline google/cloud/bigquerycontrol/CMakeLists.txt:200]: Alternative: unconditionally add the subdirectory, and return from the subdirectory if `NOT BUILD_TESTING`.

File: google/cloud/bigquerycontrol/integration_tests/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ~~~
+
+# The tests require googletest to be installed. Force CMake to use the config
+# file for googletest (that is, the CMake file installed by googletest itself),
+# because the generic `FindGTest` module does not define the GTest::gmock
+# target, and the target names are also weird.
+find_package(GTest CONFIG REQUIRED)
```
Comment (coryan) [inline google/cloud/bigquerycontrol/integration_tests/CMakeLists.txt:21]: We have a helper for this:

```
include(FindGMockWithTargets)
```

We are still a few CMake revisions away from not needed that helper:

https://cmake.org/cmake/help/latest/module/FindGTest.html#imported-targets

File: google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  minimum_year_param.set_name("minimum_year");
+  bigquery_proto::QueryParameterType minimum_year_param_type;
+  minimum_year_param_type.set_type("INT64");
+  *minimum_year_param.mutable_parameter_type() = minimum_year_param_type;
+  bigquery_proto::QueryParameterValue minimum_year_param_value;
+  auto year_value = google::protobuf::StringValue();
+  year_value.set_value("1970");
+  *minimum_year_param_value.mutable_value() = year_value;
+  *minimum_year_param.mutable_parameter_value() = minimum_year_param_value;
+  *query.add_query_parameters() = minimum_year_param;
```
Comment (coryan) [inline google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc:76]: Please use `google::protobuf::TextFormat::ParseFromString()`.  In a test, that is a more readable way to initialize complex protos with hard-coded values.

File: google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  google::cloud::bigquery::v2::ListJobsRequest list_request;
+  list_request.set_project_id(project_id_);
+  std::vector<bigquery_proto::JobReference> job_refs;
+  std::vector<std::string> job_ids;
+  for (auto job : client.ListJobs(list_request)) {
+    ASSERT_STATUS_OK(job);
+    job_refs.push_back(job->job_reference());
+    job_ids.push_back(job->job_reference().job_id());
+  }
+  EXPECT_THAT(job_ids, Contains(job_id));
```
Comment (coryan) [inline google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc:109]: Consider something like:

```cc
std::vector<...> jobs;
for (auto job : client.ListJobs(list_request)) {
  ASSERT_STATUS_OK(job);
  jobs.push_back(*std::move(job));
}
EXPECT_THAT(jobs, Contains(ResultOf("job id", [](auto const& x) { return x.job_reference().job_id(); }, job_id)));
```

That will give you (I hope) better errors.

File: google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+  google::cloud::bigquery::v2::GetJobRequest get_request;
+  get_request.set_project_id(project_id_);
+  get_request.set_job_id(job_id);
+  bool job_complete = false;
+  while (!job_complete) {
+    auto get_result = client.GetJob(get_request);
+    ASSERT_STATUS_OK(get_result);
```
Comment (coryan) [inline google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc:117]: Consider `ASSERT_THAT(get_result, IsOkAndHolds(...));`

**Review by coryan — CHANGES_REQUESTED** [review]

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14457)* status: 0 of 4 files reviewed, 5 unresolved discussions (waiting on @coryan)

___
*[`google/cloud/bigquerycontrol/CMakeLists.txt` line 200 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14457#gh-1673165483:-O1XnDQE1Ej7dBTgi8hA:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5aefd6902dbfc63c63be7a21df9279420394db46/google/cloud/bigquerycontrol/CMakeLists.txt#L200)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Alternative: unconditionally add the subdirectory, and return from the subdirectory if `NOT BUILD_TESTING`.
</blockquote></details>

Done.
___
*[`google/cloud/bigquerycontrol/integration_tests/CMakeLists.txt` line 21 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14457#gh-1673167027:-O1XnEE_04XB2xG0VdI5:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5aefd6902dbfc63c63be7a21df9279420394db46/google/cloud/bigquerycontrol/integration_tests/CMakeLists.txt#L21)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

We have a helper for this:

```
include(FindGMockWithTargets)
```

We are still a few CMake revisions away from not needed that helper:

https://cmake.org/cmake/help/latest/module/FindGTest.html#imported-targets


</blockquote></details>

Done.
___
*[`google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc` line 76 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14457#gh-1673183949:-O1XnEvD3XH0TJUqiOLg:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5aefd6902dbfc63c63be7a21df9279420394db46/google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc#L76)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Please use `google::protobuf::TextFormat::ParseFromString()`.  In a test, that is a more readable way to initialize complex protos with hard-coded values.
</blockquote></details>

Done.
___
*[`google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc` line 109 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14457#gh-1673189450:-O1XnGSw9GHIKBQ4sS0i:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5aefd6902dbfc63c63be7a21df9279420394db46/google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc#L109)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Consider something like:

```cc
std::vector<...> jobs;
for (auto job : client.ListJobs(list_request)) {
  ASSERT_STATUS_OK(job);
  jobs.push_back(*std::move(job));
}
EXPECT_THAT(jobs, Contains(ResultOf("job id", [](auto const& x) { return x.job_reference().job_id(); }, job_id)));
```

That will give you (I hope) better errors.


</blockquote></details>

Done.
___
*[`google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc` line 117 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14457#gh-1673189864:-O1XnHGX3ulKrqINSjuB:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5aefd6902dbfc63c63be7a21df9279420394db46/google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc#L117)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Consider `ASSERT_THAT(get_result, IsOkAndHolds(...));`
</blockquote></details>

Done.

**Review by coryan — APPROVED** [review]
> Reviewed 1 of 4 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14457)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @scotthart)

___
*[`google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc` line 76 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14457#-O1XoS8BELLRsY8b8PdC:-O1XoS8BELLRsY8b8PdD:brv6pjy) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/7397f7c53df5c702bbd8897adfc264484dd43371/google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc#L76)):*
> ```cpp
>   ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
>       kMinimumYearParam, &minimum_year_param));
>   *query.add_query_parameters() = minimum_year_param;
> ```

Consider



_Suggestion:_
```cpp
  auto constexpr kMinimumYearParam = R"pb(
    name: "minimum_year"
    parameter_type { type: "INT64" }
    parameter_value { value { value: "1970" } }
  )pb";  
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      kMinimumYearParam, query.add_query_parameters()));
```

File: google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+  bigquery_proto::JobConfiguration config;
+  *config.mutable_query() = query;
+  bigquery_proto::Job job;
+  *job.mutable_configuration() = config;
+  bigquery_proto::InsertJobRequest job_request;
+  job_request.set_project_id(project_id_);
+  *job_request.mutable_job() = job;
+
+  StatusOr<bigquery_proto::Job> insert_result = client.InsertJob(job_request);
```
Comment (dbolduc) [inline google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc:86]: do leaked jobs get garbage collected? If not we will need to do so.

File: google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+      jobs,
+      Contains(ResultOf(
+          "job id", [](auto const& x) { return x.job_reference().job_id(); },
+          job_id)));
+
+  google::cloud::bigquery::v2::GetJobRequest get_request;
+  get_request.set_project_id(project_id_);
+  get_request.set_job_id(job_id);
+  bool job_complete = false;
+  while (!job_complete) {
```
Comment (dbolduc) [inline google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc:117]: uhhh... can we have a counter that cuts this off after a certain number of attempts?

File: google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    auto get_result = client.GetJob(get_request);
+    ASSERT_THAT(
+        get_result,
+        testing_util::IsOkAndHolds(ResultOf(
```
Comment (dbolduc) [inline google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc:121]: `using ::google::cloud::testing_util::IsOkAndHolds;`

File: google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+  EXPECT_THAT(job_id, Not(IsEmpty()));
+  EXPECT_THAT(
+      insert_result->configuration().query().destination_table().project_id(),
+      Not(IsEmpty()));
+  EXPECT_THAT(
+      insert_result->configuration().query().destination_table().dataset_id(),
+      Not(IsEmpty()));
+  EXPECT_THAT(
+      insert_result->configuration().query().destination_table().table_id(),
+      Not(IsEmpty()));
```
Comment (dbolduc) [inline google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc:98]: optional nit: save some lines with:

```cc
auto const& table = insert_result->configuration().query().destination_table();
EXPECT_THAT(table.project_id(), ...);
EXPECT_THAT(table.dataset_id(), ...);
EXPECT_THAT(table.table_id(), ...);
```

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14457)* status: 3 of 4 files reviewed, 4 unresolved discussions (waiting on @coryan and @dbolduc)

___
*[`google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc` line 76 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14457#-O1XoS8BELLRsY8b8PdC:-O1YbWG94MMj4N8LH2t4:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/7397f7c53df5c702bbd8897adfc264484dd43371/google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc#L76)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Consider
</blockquote></details>

Done.
___
*[`google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc` line 86 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14457#gh-1674427041:-O1YamQb-MMrDFeQhLne:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/7397f7c53df5c702bbd8897adfc264484dd43371/google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc#L86)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

do leaked jobs get garbage collected? If not we will need to do so.
</blockquote></details>

Done.
___
*[`google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc` line 98 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14457#gh-1674434624:-O1Yamx4CXMgureQsmNm:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/7397f7c53df5c702bbd8897adfc264484dd43371/google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc#L98)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

optional nit: save some lines with:

```cc
auto const& table = insert_result->configuration().query().destination_table();
EXPECT_THAT(table.project_id(), ...);
EXPECT_THAT(table.dataset_id(), ...);
EXPECT_THAT(table.table_id(), ...);
```
</blockquote></details>

Done.
___
*[`google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc` line 117 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14457#gh-1674429808:-O1YanHI3BEoMx6B03mx:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/7397f7c53df5c702bbd8897adfc264484dd43371/google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc#L117)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

uhhh... can we have a counter that cuts this off after a certain number of attempts?
</blockquote></details>

Done.
___
*[`google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc` line 121 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14457#gh-1674432639:-O1YangI5M9859utHzEN:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/7397f7c53df5c702bbd8897adfc264484dd43371/google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc#L121)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

`using ::google::cloud::testing_util::IsOkAndHolds;`
</blockquote></details>

Done.

**Review by dbolduc — APPROVED** [review]

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14457)* status: 3 of 4 files reviewed, 6 unresolved discussions (waiting on @coryan and @dbolduc)

___
*[`google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc` line 115 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14457#gh-1674718440:-O1Yk9KL4wCyCBatOIbk:b-ugu45u) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a70d5fb9df13533ed88dc56c86d77d7e30a9a73e/google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc#L115)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

s/ASSERT/EXPECT/ ?
</blockquote></details>

The next line accesses the T of the StatusOr<T>, so we want to fail the test if we know that access will error.
___
*[`google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc` line 126 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14457#gh-1674718199:-O1Ykawt0kxNdgI8JXTB:b-dkemw4) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a70d5fb9df13533ed88dc56c86d77d7e30a9a73e/google/cloud/bigquerycontrol/integration_tests/job_integration_test.cc#L126)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

the loop can complete without the status being "DONE". Do we care?

`EXPECT_TRUE(job_complete)` ?
</blockquote></details>

We don't care enough to introduce potential flakes since we only wait for a limited time for the job to complete. But, if it does complete, we do have an EXPECT on deleting it which is close enough for our purposes.

**Comment by dbolduc** [comment]
> The next line accesses the T of the StatusOr, so we want to fail the test if we know that access will error.

You're right. I missed it.

**Pull request merged**